### ISSUE TITLE
Continuous active connection after apps closed and reopened

### ISSUE THREAD

**Issue by chriship** [spotify/ios-sdk#240]
After reading a lot of documentation, I am still confused about how to keep a connection with Spotify alive forever. My ideal scenario would be that a user connects with Spotify once, and then whenever they go back to my app, the connection is still active.

I've been running the SPTLoginSampleAppSwift app along with the server side token swap code but if the app is closed and reopened, I need to press the 'connect' button again. Even if Spotify is running in the bakground still.

Is it even possible to do what I want? Is the token refresh functionalty supposed to do what I want? Or is it for the scenario whereby someone will be listening to Spotify for over one hour in one sitting?

Thanks!

**Comment by kkarayannis** [comment]
> how to keep a connection with Spotify alive forever
That isn't really possible. You should focus on specific use cases where you can realistically keep the connection alive.

**Comment by chriship** [comment]
Thanks for the reply. So, all I really want is to only do the Spotify app switch once. Then, when I reopen my app, it will connect in the background (renewSession()??), Is that possible?

**Comment by kkarayannis** [comment]
It’s not possible because usually the Spotify app is suspended by iOS if it’s not playing music so it’s listening for connections. It’s not a matter if you have a valid token or not at this point. So the app switch serves as a way of waking up Spotify, in addition to authentication.

**Comment by chriship** [comment]
Right, but what about if Spotify is open, then I open my app. Can the session be renewed and not require the app switch?

**Comment by kkarayannis** [comment]
I wouldn't count on it. Spotify can shut down the socket that listens for connections immediately when it's sent to the background if it's not playing music. The only case where that might work is in split screen mode on an iPad, but in that case the app switch is not noticeable from the user's perspective so nothing is gained by avoiding it.

**Comment by chriship** [comment]
Ah ok. So the method described [here](https://github.com/spotify/ios-sdk/issues/129#issuecomment-574453868) will not work? Some users in #129 seem to have figured it out.

**Comment by logangeorge01** [comment]
I've been trying to figure this out for a few days now, and I finally have a solution. Whenever I do the app switch to connect (which ideally is only the first time opening the app), I store the refresh token in UserDefaults. Then in the applicationDidBecomeActive function in AppDelegate, I check first if it has an access token (the session is still alive or can be renewed with SPTSession) like in the Demo Projects, and if not, I retrieve the refresh token from UserDefaults and refresh it manually with a server, in turn receiving an access token which I assign to the app remote's connection parameters. Then I connect with app remote and I effectively have a permanent connection.

**Comment by fotiDim** [comment]
@logangeorge01 what is the benefit of this? If the Spotify app is not active you would still need to be redirected to Spotify before you are able to play a song.

**Comment by logangeorge01** [comment]
Yeah you're correct. There is really not much point to it if Spotify isn't active. After a month of working with different Spotify auth/streaming configurations, I've got a working setup. I use this SDK for auth, specifically the auth flow described [here](https://github.com/spotify/ios-sdk/blob/master/docs/auth.md) which doesn't require the app or start playback, and then I use the deprecated SpotifyAudioPlayback framework instead of AppRemote for streaming. This way I have the user sign in once (either by redirecting to the app or to safari), and then I store and refresh the session with the ruby script hosted on heroku every subsequent time.
